trin-execution/src/era_manager.rs

```diff
@@ -0,0 +1,69 @@
+use std::time::Duration;
+
+use e2store::{
+    era1::{BlockTuple, Era1},
+    utils::get_shuffled_era1_files,
+};
+use surf::{Client, Config};
+use tokio::time::sleep;
+use tracing::warn;
+
+pub struct EraManager {
+    current_era: Option<Era1>,
+    http_client: Client,
+    era1_files: Vec<String>,
+}
+
+impl EraManager {
+    pub async fn new() -> anyhow::Result<Self> {
+        let http_client: Client = Config::new()
+            .add_header("Content-Type", "application/xml")
+            .expect("to be able to add header")
+            .try_into()?;
+        let era1_files = get_shuffled_era1_files(&http_client).await?;
+        Ok(Self {
+            current_era: None,
+            http_client,
+            era1_files,
+        })
+    }
+
+    pub async fn get_block_by_number(&mut self, block_number: u64) -> anyhow::Result<&BlockTuple> {
+        let era1 = match &self.current_era {
+            Some(era1)
+                if era1.epoch_number() == Era1::epoch_number_from_block_number(block_number) =>
+            {
+                self.current_era.as_ref().expect("current_era to be some")
+            }
+            _ => self.current_era.insert(
+                self.download_era1_file(Era1::epoch_number_from_block_number(block_number))
+                    .await?,
+            ),
+        };
+        Ok(&era1.block_tuples[(block_number % 8192) as usize])
```

Comment:
nit: If you don't want to move `block_number % 8192` into `era1.rs` as function, at least use constant (8192) from there.